### ISSUE TITLE
If console CDN is not available, fall back to a pre-packaged version of the console?

### ISSUE THREAD

**Issue by coco98** [hasura/graphql-engine#521]
(via @thangngoc on discord)

Offline development is a bit of a pain otherwise.

@praveenweb @karthikvt26 What do you guys think about this? Pros/cons?

**Comment by karthikvt26** [comment]
It will be very useful for people developing locally using docker and postgres running locally. 

One of the major problem will be to retrieve new console changes for the same version of graphql-engine 

Toggling between offline and online can be done by configuring server to accept a flag consoleAssets = 'offline' or consoleAssets='online' to toggle between the url's?

**Comment by mwaldstein** [comment]
Suggestion on this as an alternate option - Bundle the console assets in the cli-migrations build. It is less of an effort than getting web workers implemented and gives a fairly quick solution for those of us looking for local hosting / a fully packaged version.

I feel that version is less size sensitive and is likely used in more scenarios where local hosting is desired.

**Comment by shahidhk** [comment]
@mwaldstein This should be possible. I'll give an update after hacking around a little bit.

**Comment by mnlbox** [comment]
Guys I think it's very important bug, any plan to fix this?

My team completely work offline and because of this we can't use Hasura in my development environment.
Now this is my console page as you can see it's break in `Loading...` state:
![console_error](https://user-images.githubusercontent.com/25859075/52841285-2535d200-3111-11e9-9ba5-5a7ebc12e1b8.png)

**/CC:** @shahidhk , @0x777

**Comment by patcito** [comment]
You can check the connection with navigator.onLine https://developers.google.com/web/updates/2011/06/navigator-onLine-in-Chrome-Dev-channel

It seems to be relatively well supported https://caniuse.com/#feat=online-status

```
if (navigator.onLine) {
  console.log('ONLINE!');
} else {
  console.log('Connection flaky');
}
```

**Comment by shahidhk** [comment]
Thanks @patcito, we'll bump this issue up. 

@mnlbox Till then, as a workaround, you can download `main.js`, `main.css` and `vendor.js`, keep it in a directory and load them using `hasura console --static-dir <path>` to use the downloaded assets.

**Comment by mnlbox** [comment]
@shahidhk I'm using Hasura with docker-compose is it possible trough any environment variable too?
Any short term plan to fix this?
We really need to work on project when don't have internet access too.

**Comment by shahidhk** [comment]
@mnlbox Till we fix https://github.com/hasura/graphql-engine/issues/1440, the workaround is to have the CLI installed locally and run `hasura console --static-dir <path>`

**Comment by mnlbox** [comment]
Guys please fix this before [Release Candidates Milestone](https://github.com/hasura/graphql-engine/milestone/22). Thanks for your attention :wink: 

**/CC:** @shahidhk, @dsandip

**Comment by Zoharos** [comment]
I tried using the cli in the cli-migration docker in order to set the --static-dir flag, but got this error:
`preparing execution context failed: setting up global confog failed: cannot create global config directory: mkdir/.hasura: permission denied`

**Comment by shahidhk** [comment]
@Zoharos You cannot use the CLI console from a docker image. We're working on supporting it.

But the error is something else. Your container image does not have permission to create files?

**Comment by Zoharos** [comment]
@shahidhk thanks for your answer,
what i'm trying to do is make a docker image of hasura that doesn't make request for the assets.
In order to that I understand that there is a flag that can be enabled from the console with the help of this command:
`hasura-cli console --static-dir "/static"`.
I do that from .sh file like you guys did in the cli-migrations docker file.
This error occurs when i use hasura-cli inside the container.
The base image is v1.0.0-alphav45.cli-migrations.

**Comment by shahidhk** [comment]
@Zoharos There are other issues that prevent the console form loading from within a docker image. We're ideating on multiple solutions - should have something out in a few weeks.

**Comment by Zoharos** [comment]
If there will be a solution that wouldn't require of running console command in order to lunch the console from a static directory it would be great (something like an ENV var).
Thanks, @shahidhk.

**Comment by tibotiber** [comment]
> @mnlbox Till we fix #1440, the workaround is to have the CLI installed locally and run `hasura console --static-dir <path>`

@shahidhk would you mind providing an example path, I'm trying to use this approach but can't figure it out. Thanks

**Comment by mnesarco** [comment]
Hi Friends,
Accessing the console in a local environment without internet is impossible. The `--static-dir` workaround looks more like a hack than an official solution. Is there any plan to support it just setting an env var in the near future?

Cheers,
Frank.
